### ISSUE TITLE
mod_wsgi WSGIDaemonProcess under Alpine linux causes "Segmentation fault"

### ISSUE THREAD

**Issue by sean102** [GrahamDumpleton/mod_wsgi#620]
This problem probably relates to https://groups.google.com/forum/#!topic/modwsgi/2lDoZEtAkQI

Using the WSGIDaemonProcess directive under Apache 2.4.46, Alpine linux, ProxMox Container, causes "Segmentation Fault" (Python 2.7)

After much debugging I have traced this back to a memory access after free issue.

Apache makes multiple passes over the config files. Before each pass it flushes the "pconf" memory pool in the reset_process_pconf(process) function. mod_wsgi holds references to this memory pool in the "wsgi_import_list" and "wsgi_daemon_list" globals. When apache makes its second pass over the config file, the WSGIDaemonProcess directive causes the wsgi_daemon_list global to be dereferenced but it no longer points to anything valid. Segmentation Fault ensues. 

This can be fixed by installing an apr_pool_cleanup_register() hook to NULL these pointers whenever the pool is cleared. 

Given the nature of the fault, I imaging this has been causing instabilities on all OS environments that use the Daemon mode.
I have not tested this but I suspect the WSGIScriptAlias, WSGIScriptAliasMatch, and WSGIImportScript directives might also have invoked this problem if apache did multiple passes of its config files.

I have fixed this problem on my system with the following mods. I think these should go into the main release.

mod_wsgi-4.7.1/src/server/mod_wsgi.c

--- line ~4744 ---
>         if (!wsgi_import_list) {                                             
>             wsgi_import_list = apr_array_make(cmd->pool, 20,                 
>                                               sizeof(WSGIScriptFile));
>             apr_pool_cleanup_register(cmd->pool, &wsgi_import_list,   
>                                   ap_pool_cleanup_set_null,                
>                                   apr_pool_cleanup_null);                    
>         }                                                             
> 

--- line ~5247 ---
>     if (!wsgi_import_list) {
>         wsgi_import_list = apr_array_make(cmd->pool, 20,
>                                           sizeof(WSGIScriptFile));
>         apr_pool_cleanup_register(cmd->pool, &wsgi_import_list,              
>                                   ap_pool_cleanup_set_null,                  
>                                   apr_pool_cleanup_null);             
>     }
> 

--- line ~7896 ---
>     if (!wsgi_daemon_list) {                                               
>         wsgi_daemon_list = apr_array_make(cmd->pool, 20,                     
>                                           sizeof(WSGIProcessGroup));           
>         apr_pool_cleanup_register(cmd->pool, &wsgi_daemon_list,                
>                                   ap_pool_cleanup_set_null,                
>                                   apr_pool_cleanup_null);                  
>     }                                                                         
> 

 Tested on: 
mod_wsgi 4.7.1  ("develop" branch)
Proxmox Kernel: Debian 10.5
Alpine 3.11
Python 2.7

mod_wsgi was built with the "Config Make Make-Install" method.

Tested with the following wsgi script
>     import cStringIO
>     import os
> 
>     def application(environ, start_response):
>         headers = []
>         headers.append(('Content-Type', 'text/plain'))
>         write = start_response('200 OK', headers)
> 
>         input = environ['wsgi.input']
>         output = cStringIO.StringIO()
> 
>         print >> output, "PID: %s" % os.getpid()
>         print >> output, "UID: %s" % os.getuid()
>         print >> output, "GID: %s" % os.getgid()
>         print >> output
> 
>         keys = environ.keys()
>         keys.sort()
>         for key in keys:
>             print >> output, '%s: %s' % (key, repr(environ[key]))
>         print >> output
> 
>         output.write(input.read(int(environ.get('CONTENT_LENGTH', '0'))))
> 
>         return [output.getvalue()]
>

**Comment by GrahamDumpleton** [comment]
I'd say you are pretty spot on and it is awesome you have managed to track that down. Even looks like you especially created a GitHub account just to report this issue, so even more impressed. :-)

The problem had the potential for arising when `WSGIImportScript` was used, or if `WSGIScriptAlias` was used and both the `process-group` and `application-group` options were used with that directive. The `WSGIScriptAliasMatch` directive would also have been affected if both options were used, but only if the target didn't have substitutions on the target path based on the matched URL as in that case preloading shouldn't have been done.

Even though it is an issue, seems it was extremely rare occurrence on full Linux distributions or macOS. It was only Alpine that was triggering it reliably, most likely because it doesn't use glibc and so the memory allocator behaved much differently. Given the nature of the issue, it is really surprising that it wasn't causing more problems, especially since `mod_wsgi-express` relied on using `WSGIImportScript`.

I have incorporated the fix into the `develop` branch. It should be included in version 4.8.0. I am not exactly sure when that will be as working on some other changes that want to finish first and include in that release.

Much thanks for reporting this and providing a fix.

**Comment by sean102** [comment]
Yes it is surprising that this bug wasn't more consequential. I guess if it was, it would have been picked up much earlier.

I'm not in a particular rush to get the fix into the repository. As I'm maintaining a Python 2.7 app, I have to use CMMI install anyway so I can do the patch manually. Anyone else who can't wait has everything they need above to apply the patch themselves.

Thanks for your work supporting open source software. I'm glad to be able to give something back.

**Comment by TheBirdsNest** [comment]
Good Afternoon Graham! Is this fix available anywhere yet?
I also have an issue with segmentation faults with Apache in Alpine. 


```
/etc/apache2/conf.d # httpd -X
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using 172.17.0.5. Set the 'ServerName' directive globally to suppress this message
Segmentation fault
```

**Comment by GrahamDumpleton** [comment]
Closing as was released in 4.8.0.

**Comment by TheBirdsNest** [comment]
Thanks Graham!
